walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,104 @@
+    Collection<String> meetingAttendees = request.getAttendees();
+    long meetingDuration = request.getDuration();
+
+    for (Event event : events) {
+      Set<String> eventAttendees = event.getAttendees();
+      for (String meetingAttendee : meetingAttendees) {
+        if (eventAttendees.contains(meetingAttendee)) {
+          TimeRange eventTimeRange = event.getWhen();
+          addUnavailableMinutes(unavailableMinutes, eventTimeRange);
+          break;
+        }
+      }
+    }
+    
+    Collection<TimeRange> availableTimeSlots =
+        checkForAvailableTimeSlots(unavailableMinutes, meetingDuration);
+
+    return availableTimeSlots;
+  }
+
+  /**
+   * Adds the time range of a specific event to the array of unavailable minutes.
+   * @param unavailableMinutes the minutes where other events are happening.
+   * @param eventTimeRange the time range of a specific event to be added to unavailableMinutes.
+   */
+  private void addUnavailableMinutes(boolean[] unavailableMinutes, TimeRange eventTimeRange) {
+    if (eventTimeRange.start() < 0 || eventTimeRange.end() - 1 > 24 * 60 - 1) return;
+    for (int i = eventTimeRange.start(); i <= eventTimeRange.end() - 1; i++) {
+      unavailableMinutes[i] = true;
+    }
+  }
+
+  /**
+   * Creates a collection of available time slots with the information of unavailable minutes in the
+   * day and the minimum duration required.
+   * @param unavailableMinutes the minutes during the day where there can be no meetings.
+   * @param duration the minimum duration required for the requested meeting.
+   * @return a collection of time ranges where setting a meeting is possible.
+   */
+  private Collection<TimeRange> checkForAvailableTimeSlots(
+      boolean[] unavailableMinutes, long duration) {
+    // If duration required for the meeting is longer than a day, meeting is not possible.
+    if (duration > 24 * 60) {
+      return Collections.emptyList();
+    }
+
+    List<TimeRange> availableTimeSlots = new ArrayList<>();
+    boolean inTimeRange = false;
+    int timeRangeStart = 0;
+    int timeRangeEnd = 0;
+    for (int i = 0; i < unavailableMinutes.length; i++) {
+      // If the current minute is available but not in a time range, start the time range and set
+      // this minute to the start of that time range.
+      if (!unavailableMinutes[i] && !inTimeRange) {
+        inTimeRange = true;
+        timeRangeStart = i;
+      // If the current minute is available and already in a time range, update the end of this
+      // time range to this minute.
+      } else if (!unavailableMinutes[i] && inTimeRange) {
+        timeRangeEnd = i;
+      // If the current minute is unavailable but inside a time range, end the time range and add
+      // it to the list of available time slots if its duration is greater than the required.
+      } else if (inTimeRange) {
+        inTimeRange = false;
+        if(timeRangeEnd - timeRangeStart + 1 >= duration) {
```

Comment:
Make sure there's a space between your if statement and opening paranthesis!

Reply:
Done!